### ISSUE TITLE
get_url fails with "failed to create temporary content file: (-1, 'Unexpected EOF')" when not running locally

### ISSUE THREAD

**Issue by oskapt** [ansible/ansible-modules-core#4269]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

get_url
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /home/ubuntu/ansible_mail/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

```
[defaults]
inventory = hosts
gathering = smart
host_key_checking = False
remote_user = ubuntu
ask_vault_pass = True
jinja2_extensions = jinja2.ext.do

[ssh_connection]
pipelining = True
```
##### OS / ENVIRONMENT

Ubuntu 14.04 LTS
##### SUMMARY

`get_url` works locally but fails when configuring remote host
##### STEPS TO REPRODUCE

1. Run the playbook below with `ansible-playbook -c local test.yml` and it will work
2. Run it again connecting to a remote system and it will fail

I have the same behavior on multiple hosts. I turned off pipelining to see if it mattered, and it didn't.

```

---
- hosts: host.foo.com
  gather_facts: False
  vars:
    - tarsnap_version: 1.0.35
    - sha256sum: 6c9f6756bc43bc225b842f7e3a0ec7204e0cf606e10559d27704e1cc33098c9a
  tasks:
    - name: Download Tarsnap source
      get_url:
        url=https://www.tarsnap.com/download/tarsnap-autoconf-{{ tarsnap_version }}.tgz
        dest=/var/tmp/tarsnap-autoconf-{{ tarsnap_version }}.tgz
        sha256sum={{ sha256sum }}
```

##### EXPECTED RESULTS

I expected that the tarsnap source would be downloaded.
##### ACTUAL RESULTS

It fails with "failed to create temporary content file: (-1, 'Unexpected EOF')"

```
ubuntu@docker01:~/ansible_mail$ ansible-playbook -vvvv test.yml 
Using /home/ubuntu/ansible_mail/ansible.cfg as config file
Vault password: 
Loaded callback default of type stdout, v2.0

PLAYBOOK: test.yml *************************************************************
1 plays in test.yml

PLAY [host.foo.com] ********************************************************

TASK [Download Tarsnap source] *************************************************
task path: /home/ubuntu/ansible_mail/test.yml:8
<31.171.245.146> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<31.171.245.146> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o ControlPath=/home/ubuntu/.ansible/cp/ansible-ssh-%h-%p-%r 31.171.245.146 '/bin/sh -c '"'"'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python && sleep 0'"'"''
fatal: [host.foo.com]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"backup": false, "checksum": "", "content": null, "delimiter": null, "dest": "/var/tmp/tarsnap-autoconf-1.0.35.tgz", "directory_mode": null, "follow": false, "force": false, "force_basic_auth": false, "group": null, "headers": null, "http_agent": "ansible-httpget", "mode": null, "owner": null, "regexp": null, "remote_src": null, "selevel": null, "serole": null, "setype": null, "seuser": null, "sha256sum": "6c9f6756bc43bc225b842f7e3a0ec7204e0cf606e10559d27704e1cc33098c9a", "src": null, "timeout": 10, "tmp_dest": "", "url": "https://www.tarsnap.com/download/tarsnap-autoconf-1.0.35.tgz", "url_password": null, "url_username": null, "use_proxy": true, "validate_certs": true}, "module_name": "get_url"}, "msg": "failed to create temporary content file: (-1, 'Unexpected EOF')"}

NO MORE HOSTS LEFT *************************************************************
        to retry, use: --limit @test.retry

PLAY RECAP *********************************************************************
host.foo.com           : ok=0    changed=0    unreachable=0    failed=1   

```

**Comment by ansibot** [comment]
@jpmens ping, this issue is waiting for your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jpmens** [comment]
Cannot and will not look at this until end of August earliest.

**Comment by oskapt** [comment]
That's fine. I am working around it with `shell: curl -O` for now. Get to it when you can. :smile:

**Comment by zachmullen** [comment]
I have the same issue. I notice it is happening when the server returns a 302 redirect.

**Comment by ansibot** [comment]
@jpmens, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jpmens** [comment]
needs_contributor

**Comment by naterkane** [comment]
i'm having the same issue as well. 

```
- name: Download WordPress
  get_url:
    url: https://wordpress.org/wordpress-{{ wp_version }}.tar.gz
    dest: /tmp/ #wordpress-{{ wp_version }}.tar.gz
    checksum: sha1:{{ wp_sha1sum }}
```

```
TASK [blog : Download WordPress] ***********************************************
task path: /Users/naterkane/Projects/projectname/ansible/roles/blog/tasks/main.yml:23
<localhost.projectname.io> ESTABLISH CONNECTION FOR USER: root on PORT 22 TO localhost.tenon.io
<localhost.projectname.io> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1473580022.13-185217142696645 `" && echo ansible-tmp-1473580022.13-185217142696645=
"` echo $HOME/.ansible/tmp/ansible-tmp-1473580022.13-185217142696645 `" ) && sleep 0'
<localhost.projectname.io> PUT /var/folders/nh/sv3z8c0s6gj23ssfzxt10fnh0000gn/T/tmpdIcKYg TO /root/.ansible/tmp/ansible-tmp-1473580022.13-185217142696645/get_url
<localhost.projectname.io> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1473580022.13-185217142696645/get_url;
 rm -rf "/root/.ansible/tmp/ansible-tmp-1473580022.13-185217142696645/" > /dev/null 2>&1 && sleep 0'
fatal: [localhost.projectname.io]: FAILED! => {
  "changed": false,
  "failed": true,
  "invocation": {
    "module_args": {
      "backup": false,
      "checksum": "sha1:d3a70d0f116e6afea5b850f793a81a97d2115039",
      "content": null,
      "delimiter": null,
      "dest": "/tmp/",
      "directory_mode": null,
      "follow": false,
      "force": false,
      "force_basic_auth": false,
      "group": null,
      "headers": null,
      "http_agent": "ansible-httpget",
      "mode": null,
      "owner": null,
      "regexp": null,
      "remote_src": null,
      "selevel": null,
      "serole": null,
      "setype": null,
      "seuser": null,
      "sha256sum": "",
      "src": null,
      "timeout": 10, 
      "tmp_dest": "", 
      "url": "https://wordpress.org/wordpress-4.2.2.tar.gz", 
      "url_password": null, 
      "url_username": null, 
      "use_proxy": true, 
      "validate_certs": true
    }, 
    "module_name": "get_url "
  }, 
  "msg": "failed to create temporary content file: (-1, 'Unexpected EOF')"
}

```

**Comment by MartijnBraam** [comment]
I have noticed that this only happens on Ubuntu 14.04. I've run the same steps on debian 7, debian 8 and ubuntu 16.04 and it run's without problem. It fails on a fresh installed Ubuntu 14.04 VM.

**Comment by henderbj** [comment]
I was having same problem. I use sha256 checksum, but it fails when using "sha256sum" parameter, and it works when using "checksum" parameter. I added "tmp_dest" parameter too.
My changes were:

```
-    sha256sum: "{{ sha256sum }}"
+    checksum: "sha256:{{sha256sum}}"
+    tmp_dest: /tmp/
```

I am using Ubuntu 14.04 too. The default value for tmp_dest is TMPDIR, TEMP or TMP environment variables, but they seem to not be defined on Ubuntu 14.04.

A fix should be make tmp_dest parameter to default to /tmp if all of TMPDIR, TEMP or TMP are undefined.

**Comment by Mykolaichenko** [comment]
Its new issue, playbook still working on old version of ansible 
As workaround i've replaced get_url by wget

**Comment by henderbj** [comment]
I did solve on my playbook, as indicated on Oct 3rd.

**Comment by masteinhauser** [comment]
Seeing this same issue still in Ansible 2.2. I haven't attempted yet with the test case first shared, but I tried to use the `checksum` and `tmp_dest` fixes without success. 

I'll circle back on the initial test case in the morning and report back. I'll also likely move us to using wget to unblock us in the meantime.

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by oskapt** [comment]
This issue was moved to ansible/ansible#18894

**Comment by Alex2357** [comment]
I have the same problem in 2.2. Initially I implemented and it was working, but I connected to localhost to run playbook. Right now tried few times to run against another machine and it did not work.
